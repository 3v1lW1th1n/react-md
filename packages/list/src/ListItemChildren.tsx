import React, { FC, ReactNode, Fragment } from "react";

import ListItemIcon, { ListItemIconPosition } from "./ListItemIcon";
import ListItemText from "./ListItemText";

export interface ListItemChildrenProps {
  children?: ReactNode;
  /**
   * An optional className to apply to the `<span>` that surrounds the `primaryText` and optionally
   * `secondaryText` within the list item.
   */
  textClassName?: string;

  /**
   * An optional className to apply to the `<span>` that surrounds the `secondaryText` within the
   * list item.
   */
  secondaryTextClassName?: string;

  /**
   * Boolean if the children should be treated as the `primaryText` prop. This will wrap them in
   * an additional class so that they have ellipsis for text overflow.
   *
   * If you want to have more "freedom" within the ListItem, you can disable this prop so that the
   * height will grow depending on content.
   *
   * NOTE: If the `secondaryText` prop is provided, this will always be considered `true`.
   */
  textChildren?: boolean;

  /**
   * An optional element that should be rendered as the `primaryText` within the list item. It is
   * most likely easier to use the `children` prop instead, but this allows you to create more\
   * complex components with the ListItem since you can provided `children` and have the styles for
   * the `primaryText` still applied. By default, this will only allow one line of text and add
   * ellipsis for any text overflow.
   */
  primaryText?: ReactNode;

  /**
   * An optional element that should be rendered as the `secondaryText` within the list item. By
   * default, this will only span one line and add ellipsis for overflow.
   */
  secondaryText?: ReactNode;

  /**
   * An optional icon to display to the left of the children or provided text elements. If this is
   * a valid React element, the spacing class names will be cloned into the element. Otherwise it
   * will be wrapped with a `<span>` to have the correct class name applied. You can also use the
   * `forceIconWrap` prop to **always** wrap the icon in a `<span>` with the correct class name
   * applied.
   */
  leftIcon?: ReactNode;

  /**
   * An optional icon to display to the right of the children or provided text elements. If this is
   * a valid React element, the spacing class names will be cloned into the element. Otherwise it
   * will be wrapped with a `<span>` to have the correct class name applied. You can also use the
   * `forceIconWrap` prop to **always** wrap the icon in a `<span>` with the correct class name
   * applied.
   */
  rightIcon?: ReactNode;

  /**
   * Boolean if the left and/or right icons should be "forcefully" wrapped in a `<span>` with the
   * spacing class names applied instead of attempting to clone it into the provided icon element.
   */
  forceIconWrap?: boolean;

  /**
   * An optional avatar to display to the left of the children or provided text elements. If this is
   * a valid React element, the spacing class names will be cloned into the element. Otherwise it
   * will be wrapped with a `<span>` to have the correct class name applied. You can also use the
   * `forceIconWrap` prop to **always** wrap the icon in a `<span>` with the correct class name
   * applied.
   */
  leftAvatar?: ReactNode;

  /**
   * An optional avatar to display to the right of the children or provided text elements. If this is
   * a valid React element, the spacing class names will be cloned into the element. Otherwise it
   * will be wrapped with a `<span>` to have the correct class name applied. You can also use the
   * `forceIconWrap` prop to **always** wrap the icon in a `<span>` with the correct class name
   * applied.
   */
  rightAvatar?: ReactNode;

  /**
   * An optional graphic to be placed left of the main content.
   */
  leftMedia?: ReactNode;
  leftMediaLarge?: ReactNode;
  rightMedia?: ReactNode;
  rightMediaLarge?: ReactNode;

  leftPosition?: ListItemIconPosition;
  rightPosition?: ListItemIconPosition;

  /**
   * Boolean if the color pollution should be fixed at this level. This is really used when
   * composing the children with clickable list items and tree items.
   */
}

type DefaultProps = Required<
  Pick<ListItemChildrenProps, "leftPosition" | "rightPosition">
>;
type WithDefaultProps = ListItemChildrenProps & DefaultProps;

const ListItemChildren: FC<ListItemChildrenProps> = props => {
  const {
    textClassName,
    secondaryTextClassName,
    textChildren,
    primaryText,
    secondaryText,
    leftIcon,
    leftAvatar,
    leftMedia,
    leftMediaLarge,
    leftPosition,
    rightIcon,
    rightAvatar,
    rightMedia,
    rightMediaLarge,
    rightPosition,
    forceIconWrap,
    children: propChildren,
  } = props as WithDefaultProps;
  const stringifiedChildren =
    typeof propChildren === "number" ? `${propChildren}` : propChildren;
  let children = stringifiedChildren;
  if (primaryText || secondaryText || textChildren) {
    children = (
      <ListItemText
        className={textClassName}
        secondaryText={secondaryText}
        secondaryTextClassName={secondaryTextClassName}
      >
        {(textChildren && children) || primaryText}
      </ListItemText>
    );
  }

  children = (
    <ListItemIcon
      avatar={!!leftAvatar}
      media={!!leftMedia}
      mediaLarge={!!leftMediaLarge}
      icon={leftIcon || leftAvatar || leftMedia || leftMediaLarge}
      forceIconWrap={forceIconWrap}
      before
      position={leftPosition}
    >
      {children}
    </ListItemIcon>
  );
  children = (
    <ListItemIcon
      avatar={!!rightAvatar}
      media={!!rightMedia}
      mediaLarge={!!rightMediaLarge}
      icon={rightIcon || rightAvatar || rightMedia || rightMediaLarge}
      forceIconWrap={forceIconWrap}
      before={false}
      position={rightPosition}
    >
      {children}
    </ListItemIcon>
  );

  return (
    <Fragment>
      {children}
      {(primaryText && stringifiedChildren) || null}
    </Fragment>
  );
};

const defaultProps: DefaultProps = {
  leftPosition: "middle",
  rightPosition: "middle",
};

ListItemChildren.defaultProps = defaultProps;

export default ListItemChildren;
