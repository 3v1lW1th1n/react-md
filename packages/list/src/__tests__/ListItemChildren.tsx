import React from "react";
import { render } from "@testing-library/react";

import ListItemChildren from "../ListItemChildren";

describe("ListItemChildren", () => {
  it("should render correctly", () => {
    const props = { children: "Content" };
    const { container, rerender } = render(<ListItemChildren {...props} />);

    expect(container).toMatchSnapshot();

    rerender(<ListItemChildren {...props} textChildren />);
    expect(container).toMatchSnapshot();

    rerender(<ListItemChildren primaryText="Content" />);
    expect(container).toMatchSnapshot();

    rerender(
      <ListItemChildren primaryText="Primary" secondaryText="Secondary" />
    );
    expect(container).toMatchSnapshot();
  });

  it("should be able to render content to the left of the children where only one of the left* props will be used", () => {
    const props = { primaryText: "Primary Text" };
    const leftIcon = <span data-testid="left-icon">Left Icon</span>;
    const leftAvatar = <span data-testid="left-avatar">Left Avatar</span>;
    const leftMedia = <span data-testid="left-media">Left Media</span>;
    const leftMediaLarge = (
      <span data-testid="left-media-large">Left Media Large</span>
    );

    const { container, rerender } = render(
      <ListItemChildren {...props} leftIcon={leftIcon} />
    );

    expect(container).toMatchSnapshot();

    rerender(<ListItemChildren {...props} leftAvatar={leftAvatar} />);
    expect(container).toMatchSnapshot();

    rerender(<ListItemChildren {...props} leftMedia={leftMedia} />);
    expect(container).toMatchSnapshot();

    rerender(<ListItemChildren {...props} leftMediaLarge={leftMediaLarge} />);
    expect(container).toMatchSnapshot();

    rerender(
      <ListItemChildren
        {...props}
        leftIcon={leftIcon}
        leftAvatar={leftAvatar}
        leftMedia={leftMedia}
        leftMediaLarge={leftMediaLarge}
      />
    );
    expect(container).toMatchSnapshot();

    rerender(
      <ListItemChildren
        {...props}
        leftAvatar={leftAvatar}
        leftMedia={leftMedia}
        leftMediaLarge={leftMediaLarge}
      />
    );
    expect(container).toMatchSnapshot();

    rerender(
      <ListItemChildren
        {...props}
        leftMedia={leftMedia}
        leftMediaLarge={leftMediaLarge}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
