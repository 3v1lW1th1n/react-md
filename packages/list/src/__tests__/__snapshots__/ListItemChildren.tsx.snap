// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItemChildren should be able to render content to the left of the children where only one of the left* props will be used 1`] = `
<div>
  <span
    class="rmd-icon--before rmd-list-item__icon rmd-list-item__icon--before"
    data-testid="left-icon"
  >
    Left Icon
  </span>
  <span
    class="rmd-list-item__text"
  >
    Primary Text
  </span>
</div>
`;

exports[`ListItemChildren should be able to render content to the left of the children where only one of the left* props will be used 2`] = `
<div>
  <span
    class="rmd-icon--before rmd-list-item__icon rmd-list-item__icon--before rmd-list-item__icon--avatar-before"
    data-testid="left-avatar"
  >
    Left Avatar
  </span>
  <span
    class="rmd-list-item__text"
  >
    Primary Text
  </span>
</div>
`;

exports[`ListItemChildren should be able to render content to the left of the children where only one of the left* props will be used 3`] = `
<div>
  <span
    class="rmd-text-icon-spacing rmd-icon--before rmd-list-item__icon rmd-list-item__icon--before rmd-list-item__icon--media"
  >
    <span
      data-testid="left-media"
    >
      Left Media
    </span>
  </span>
  <span
    class="rmd-list-item__text"
  >
    Primary Text
  </span>
</div>
`;

exports[`ListItemChildren should be able to render content to the left of the children where only one of the left* props will be used 4`] = `
<div>
  <span
    class="rmd-icon--before rmd-list-item__icon rmd-list-item__icon--before rmd-list-item__icon--media rmd-list-item__icon--media-large"
    data-testid="left-media-large"
  >
    Left Media Large
  </span>
  <span
    class="rmd-list-item__text"
  >
    Primary Text
  </span>
</div>
`;

exports[`ListItemChildren should be able to render content to the left of the children where only one of the left* props will be used 5`] = `
<div>
  <span
    class="rmd-text-icon-spacing rmd-icon--before rmd-list-item__icon rmd-list-item__icon--before rmd-list-item__icon--avatar-before rmd-list-item__icon--media rmd-list-item__icon--media-large"
  >
    <span
      data-testid="left-icon"
    >
      Left Icon
    </span>
  </span>
  <span
    class="rmd-list-item__text"
  >
    Primary Text
  </span>
</div>
`;

exports[`ListItemChildren should be able to render content to the left of the children where only one of the left* props will be used 6`] = `
<div>
  <span
    class="rmd-text-icon-spacing rmd-icon--before rmd-list-item__icon rmd-list-item__icon--before rmd-list-item__icon--avatar-before rmd-list-item__icon--media rmd-list-item__icon--media-large"
  >
    <span
      data-testid="left-avatar"
    >
      Left Avatar
    </span>
  </span>
  <span
    class="rmd-list-item__text"
  >
    Primary Text
  </span>
</div>
`;

exports[`ListItemChildren should be able to render content to the left of the children where only one of the left* props will be used 7`] = `
<div>
  <span
    class="rmd-text-icon-spacing rmd-icon--before rmd-list-item__icon rmd-list-item__icon--before rmd-list-item__icon--media rmd-list-item__icon--media-large"
  >
    <span
      data-testid="left-media"
    >
      Left Media
    </span>
  </span>
  <span
    class="rmd-list-item__text"
  >
    Primary Text
  </span>
</div>
`;

exports[`ListItemChildren should render correctly 1`] = `
<div>
  Content
</div>
`;

exports[`ListItemChildren should render correctly 2`] = `
<div>
  <span
    class="rmd-list-item__text"
  >
    Content
  </span>
</div>
`;

exports[`ListItemChildren should render correctly 3`] = `
<div>
  <span
    class="rmd-list-item__text"
  >
    Content
  </span>
</div>
`;

exports[`ListItemChildren should render correctly 4`] = `
<div>
  <span
    class="rmd-list-item__text"
  >
    Primary
    <span
      class="rmd-list-item__text rmd-list-item__text--secondary"
    >
      Secondary
    </span>
  </span>
</div>
`;
